### ISSUE TITLE
Live assist

### ISSUE THREAD

**Issue by Cysioland** [LibreTime/libretime#480]
That's a big one, basically, one that is a deal-breaker for many stations, and one that was requested for original Airtime for years. If we could ditch the commercial software for LibreTime, that'd be great.

Basically: sound comes out of Airtime through ALSA/JACK/PulseAudio/whatever (this is currently supported), comes onto a mixing board, gets mixed with microphones and maybe a sampler, and goes on air.

There's a problem with this: you can either playback the music through LibreTime (but then have no support for playing live audio), or go on air with live audio (but then you have to use an external music player).

A dirty workaround for this is streaming your mix through Icecast's [insert your favorite sound server] input feature. But then you don't get metadata and those fancy features LibreTime does with their streams (listening stats, etc.)

I know a bit of PHP, and could provide some help. I also have access to a small Internet radio that uses hardware mixer workflow, in case we'd need to test something.

**Comment by Robbt** [comment]
Indeed, this functionality isn't currently offered. Mixxx has a lot of this functionality but it doesn't integrate with the Library. One useful feature would be the ability to just dial in during a break (something LibreTime makes difficult because playback timing etc.

**Comment by JohnnyC1951** [comment]
Having spent a week in the LiquidSoap mire, it seems that the reason that the alsa jack doesn't work is that it is not compiled into LiquidSoap although it should be possible. Many people (inc. me) cannot successfully compile without omitting Alsa/Jack and AAC. It does seem that LS is stalled and although I heard glimmers of hope a few weeks back, seems to remain so. Some of the required sources are are too old and will no longer compile in Opam or Occam. Some weeks back, I did suggest using something other than LS in the longer term and got slapped down. I do know that this functionality is possible and have one station that uses gStreamer instead of LS at its core along with a bunch of java libraries. Longer term we do need a solution to all this,

**Comment by hairmare** [comment]
Most of the things I've been considering for live assist went down the external player/mixing-desk route. I was after sharing the audio database, maybe using LibreTime as routing solution and, getting the chain to pass PAD data properly.

There is [a fork of IDJC](https://sourceforge.net/u/bmillham/idjc/ci/feature/postgres-with-plugins/~/tree/) that has a branch with support for accessing an Airtime database directly (and then going the external player route while side-loading the actual audio files over a network mount).

It was done and we used it when we where in the PoC concept phase of evaluating Airtime. It's likely that it could be made to work with LibreTime but I'm not at the point where I would recommend the set-up since we never got to qualifying it for 24/7 at our station.

I'm not sure if MIxxx supports anything integrate-able with LibreTime. The original Airtime Mixxx Skin was broken at some point as far as I can remember.

Currently we are also considering Rivendell at our station. If we get to using Rivendell for live assist our setup will most likely be rather specific to our AoIP network and it's core audio routing capabilities.

**Comment by Robbt** [comment]
One idea I had regarding this would be the direct integration of a microphone/audio recording interface into the LibreTime UI. It would seem like the mixed latency might be an issue with this because there would be no way to sync completely with what was playing on the stream. But a basic workflow that many people doing music shows use is to record short segments in between songs and upload them as seperate mp3 files and then arrange them. This works for pre-recording but opening audacity or another recording interface and then exporting an mp3 and uploading it a somewhat cumbersome experience vs. the idea of a recording interface directly in libretime that could access the computers sound-input and record it. I haven't researched it fully, I know this functionality was often times done via Flash and is actually easy enough with HTML5 and javascript see - https://stackoverflow.com/questions/16413063/html5-record-audio-to-file - so in theory something like this could be done as a rudimentary workaround.

The basic idea is you could record short intros/outros on the fly via the web UI and then slot them in between tracks. Now I think this could work with the current paradigm of playback of pre-recorded files. Actually mixing in or piping in live from the microphone w/o using the icecast stream override would likely involve some complex liquidsoap maneuvers that I can't even pretend to have spent time attempting to wrangle with. But looking it up there does appear to be the ability to add two inputs together in liquidsoap see (http://www.liquidsoap.info/doc-1.3.3/reference.html#add) - which in theory could be used to overlay a microphone input from the webpage (or another input) w/ whatever was already playing. This would be cool and really make LibreTime more useful to use for people who want to give their DJs the ability to jump in w/o completely pulling the plug on whatever was already playing.

The tricky part would be ensuring the DJ could listen to what was being played in real time w/o delay because even the playback is somewhat buffered via icecast and the alsa out is buffered 30 seconds or so. But those are my thoughts on this.

**Comment by nikmartin** [comment]
@Robbt webrtc is also an option, as user's browser->webrtc recording interface->Liquidsoap is already supported: https://github.com/webcast/webcaster is small test harness showing how to stream realtime audio from user browser (and whatever mic they have connected) to a server via websockets/webrtc

**Comment by JohnnyC1951** [comment]
Been there. Done that. Lots of people tried, WebRTC is opus.
LS doesn't like OPUS much.
It needs a webtc LT client.
Some icons gave up. Loks @asantoni

**Comment by frecuencialibre** [comment]
https://github.com/AzuraCast/AzuraCast/issues/969
